src/gluonnlp/models/t5.py

```diff
@@ -0,0 +1,849 @@
+        return hidden_states
+
+
+@use_np
+class T5Stack(HybridBlock): 
+    def __init__(
+        self, 
+        d_model, 
+        d_kv, 
+        d_ff, 
+        is_decoder, 
+        num_layers=12, 
+        num_heads=12, 
+        dropout_prob=0.1, 
+        layer_norm_eps=1E-6, 
+        activation='relu', 
+        init_factor=1.0, 
+        layout='NT', 
+        dtype='float32'
+    ): 
+        super().__init__()
+        self._is_decoder = is_decoder
+        self._d_model = d_model
+        self._d_kv = d_kv
+        self._d_ff = d_ff
+        self._num_layers = num_layers
+        self._num_heads = num_heads
+        self._inner_dim = num_heads * d_kv
+        self._dtype = dtype
+        assert layout in ['TN', 'NT'], \
+            'Invalid layout: {}. Only "TN" and "NT" are supported.'.format(layout)
+        self._layout = layout
+
+        self.relative_position_encoder = RelAttentionScoreCell(
+            query_units=self._inner_dim, 
+            num_heads=num_heads, 
+            method='t5', 
+            bidirectional=(not is_decoder), 
+            embed_initializer=Normal(d_model ** -0.5 * init_factor), 
+            layout='NTK' if layout == 'NT' else 'TNK', 
+            dtype=dtype
+        )
+        self.layers = nn.HybridSequential()
+        for _ in range(num_layers): 
+            self.layers.add(
+                T5Block(
+                    d_model=d_model, 
+                    d_kv=d_kv, 
+                    d_ff=d_ff, 
+                    is_decoder=is_decoder, 
+                    num_heads=num_heads, 
+                    dropout_prob=dropout_prob, 
+                    layer_norm_eps=layer_norm_eps, 
+                    activation=activation, 
+                    init_factor=init_factor, 
+                    layout=layout, 
+                    dtype=dtype
+                )
+            )
+        self.final_layer_norm = T5LayerNorm(
+            d_model=d_model, 
+            epsilon=layer_norm_eps, 
+            init_factor=init_factor, 
+            dtype=dtype
+        )
+        self.dropout = nn.Dropout(dropout_prob)
+
+    @property
+    def layout(self): 
+        return self._layout
+
+    @_assert_decoder_method
+    @property
+    def state_batch_axis(self): 
+        return list(layer.state_batch_axis for layer in self.layers)
+
+    @_assert_decoder_method
+    def init_states(self, batch_size, ctx, dtype='float32'): 
+        return list(layer.init_states(batch_size, ctx, dtype) for layer in self.layers)
+
+    @_assert_decoder_method
+    def incremental_decode(
+        hidden_states, 
+        past_key_value, 
+        mem_states, 
+        mem_valid_length
+    ): 
+        raise NotImplementedError
+
+    def _get_query_key_shape(self, hidden_states, mem_states, past_key_value): 
+        # NT: (B, L_seq, inner_dim); TN: (L_seq, B, inner_dim)
+        index = 1 if self.layout == 'NT' else 0
+        query_length = hidden_states.shape[index]
```

Comment:
This will make it not be able to hybridize.

Reply:
It should be able to under deferred compute, provided that it doesn't change in the hybridized model. cc @leezu to confirm

Reply:
The length of the key and query may change and I think it cannot deal with that even with deferred compute?

Reply:
@szha If that’s the case, it should be advertised. For me, I feel that the final symbol will be the same even if we have changed query_length and key_length.

Reply:
In deferred compute, the shape information is accessible as shape inference happens eagerly. If that value gets passed into a node attribute, then that attribute is frozen in the hybridized graph.

For example, suppose the query length is then used in normalization `x/length`, the operator invoked would be a broadcast scalar division, where the node attribute value of the divisor is from length, a python integer from `.shape[index]`. In that case, as the length changes, the division value would be wrong. To address it, one can use `.shape_array[index]` so that the op invoked is broadcast division and it's always the up-to-date length value that will be used.

Reply:
The current problem is that we will need to use query_length, key_length to create new matrices. For example, mx.np.arange(query_length).

Reply:
Is this also supported? I think it should not be runnable.

Reply:
Not yet. It will be possible once we get to creation ops with symbolic shapes. (cc @barry-jin about the example use case above)

Reply:
> I think it should not be runnable.

A error will be raised once you input shapes that would yield a different intermediate matrix

Reply:
> > I think it should not be runnable.
> 
> A error will be raised once you input shapes that would yield a incompatible intermediate matrix

Yes, this simply indicates that this feature is not supported yet and we should not recommend users to use `.shape` in forward.